### ISSUE TITLE
Support the `@private` JSDoc directive to exclude private APIs from type declarations

### ISSUE THREAD

**Issue by SethFalco** [microsoft/TypeScript#61651]
### 🔍 Search Terms

"private jsdoc"

### ✅ Viability Checklist

- [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [x] This wouldn't change the runtime behavior of existing JavaScript code
- [x] This could be implemented without emitting different JS based on the types of the expressions
- [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [x] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [x] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

Could properties, functions, and classes annotated with the [`@private`](https://jsdoc.app/) JSDoc directive be omitted from the emitted declarations?

tsc should throw an error if a symbol is marked as `@private`, but is actually imported/exported from another file or used in the signature of another symbol like a function parameter or return type. (And therefore isn't actually an internal API.)

I'm wary this may break some projects if implemented without an opt-in, so it may require an option to enable this behavior, to preserve backward compatibility.

### 📃 Motivating Example

I maintain a [svgo](https://github.com/svg/svgo), a library that is written in JavaScript and typed with JSDoc directives, and uses tsc to emit declaration files.

As a library normally has an intended public API, afaik there's no need to distribute type declarations for the internal API.

### 💻 Use Cases

This is just to reduce the amount of type declarations served to developers if they aren't needed, reducing the size of the package on npm a little.

No workaround is currently needed, assuming the project is a module that uses `exports` rather than `main`, as we can limit the public API ourselves. The extra types do not cause any harm, it just makes the final package a little bigger.

**Comment by SethFalco** [comment]
Thank you for providing advice on the matter! I personally would prefer if I could stick to standard JSDoc directives like `@private`, but willing to use `@internal` for now if that can achieve this. 

However, I've just tried this but when I add `@internal` and enable `stripInternal: true`, the function is not removed from the emitted type declaration files. 🤔

I'm using tsc v5.8.3, do you have any idea what could be wrong here? 🤔

**`css-select-adapter.js` (snippet from input)**
```js
/**
 * @param {Map<import('../types.js').XastNode, import('../types.js').XastParent>} parents
 * @returns {Required<import('css-select').Options<import('../types.js').XastNode & { children?: any }, import('../types.js').XastElement>>['adapter']}
 * @internal
 */
export function createAdapter(parents) {
```

**`tsconfig.build.json` (config)**
```json
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "noEmit": false,
    "declaration": true,
    "emitDeclarationOnly": true,
    "outDir": "types/",
    "stripInternal": true,
  },
  "include": [
    "lib/svgo-node.js",
    "lib/svgo.js"
  ]
}
```

**`css-select-adapter.d.ts` (output)**
```ts
/**
 * @internal
 * @param {Map<import('../types.js').XastNode, import('../types.js').XastParent>} parents
 * @returns {Required<import('css-select').Options<import('../types.js').XastNode & { children?: any }, import('../types.js').XastElement>>['adapter']}
 */
export function createAdapter(parents: Map<import("../types.js").XastNode, import("../types.js").XastParent>): Required<import("css-select").Options<import("../types.js").XastNode & {
    children?: any;
}, import("../types.js").XastElement>>["adapter"];
```
